KAFKA-5166: Add option "dry run" to Streams application reset tool

bharatviswa504 opened the pull request:
Addressed the below review comment from #PR #2998 from @mjsax 

I am wondering if it would be better, to "embed" the dry-run into the actual code and branch on each place. Otherwise, if things get changed, we could easily introduce bugs (ie, dry run show something different than what the actual reset code does.

We could introduce methods like mabyeSeekToBeginning() that either does the seek or only prints to stdout. This would ensure that the main logic is used to "feed" into dry-run and we don't have code duplication.

WDYT?

Commit:
KAFKA-5166: Add option dry run to Streams application reset tool

Commit:
Updated to add consumer group in description

Commit:
addressed review comment to add dry-run logic with main code

bharatviswa504 commented:
@mjsax I initially thought it will be good. But I thought the code will not look cleaner it will have a lot of if else, for simple logic. But as you said, if main logic changes, if someone missed to update dryRun it will deviate from actual behavior. Update the code accordingly.

Commit:
checkstyle errors

mjsax commented:
I just realized that the JavaDoc of this class is not correct: Can you please fix this as a side task?

It says: "setting the application's consumer offsets for input and internal topics to zero". This should be "setting the application's consumer offsets for input topics to the beginning of the log". And we should also add to bullet point about "deleted topic" that the corresponding offsets will be deleted automatically.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
+                return;
+            } else {
+                if (inputTopics.size() != 0) {
+                    System.out.println("Resetting offsets to zero for input topics " + inputTopics + " and all internal topics.");
```
Can you please fix this output too -- the tool does "seek to beginning" and does not set offsets to zero.

bharatviswa504 replied:
fixed.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+                }
             }
-            client.commitSync();
+            client.close();
```
Not required. Client will be automatically closes, as we use Java's "try with resource" feature.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
+                            "--delete", "--topic", topic});
+                        TopicCommand.deleteTopic(zkUtils, commandOptions);
+                    } else {
+                        System.out.println("Topic - " + topic);
```
Nit: maybe (`"Topic: " + topic)` ?

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
         }
 
-        System.out.println("Done.");
+        if (!dryRun) {
```
Do we need to suppress this in dry run?

bharatviswa504 replied:
Just added to make the output look clear.

As if it is not that major, removed the check to print done.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
-    private void deleteInternalTopics(final ZkUtils zkUtils) {
-        System.out.println("Deleting all internal/auto-created topics for application " + options.valueOf(applicationIdOption));
+    private void maybedeleteInternalTopics(final ZkUtils zkUtils) {
+        Boolean dryRun = options.has(dryRunOption);
```
nit: add `final`
Why not use `boolean` ?

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
+        if (!dryRun) {
+            System.out.println("Deleting all internal/auto-created topics for application " + options.valueOf(applicationIdOption));
+        } else {
+            System.out.println("Following internal/auto-created topics will be Deleted for application " + options.valueOf(applicationIdOption));
```
I think this message is not much different from the non-dry-run one. Maybe we can have one "unified" message that can be used for both cases (to reduce the number of `if-else` branches?

bharatviswa504 replied:
Added the check because, in dry-run we only print actions, and not do the action.
As this statement looks like Deleting all internal/auto-created topics for application g1 is like it is an action.

mjsax replied:
I understand the motivation and agree it's a good idea. It's only that 

`"Deleting all internal/auto-created topics for application <app.id>."`
and
`"Following internal/auto-created topics will be Deleted for application <app.id>."`

sound super similar. So not sure if we gain much. But it's quite subjective. It's ok if you leave as is. Please fix the typo though `Deleted` -> `deleted`.

bharatviswa504 replied:
Agreed. Deleted the condition, as anyway there is this displayed on start, when run with dry-run option.
----Dry run displays the actions which will be performed when running Streams Reset Tool----

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+
+    }
+
+    private void maybeSeekToBeginning(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> inputAndInternalTopicPartitions, List<String> internalTopics) {
```
nit: add `final` to all parameters -- also, the line is quite long, maybe start a new line for each parameter?

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+            if (!dryRun) {
+                client.seekToBeginning(inputAndInternalTopicPartitions);
+            } else {
+                System.out.println("Following input topics offsets will be reset to beginning (for consumer group " + groupId + ")");
```
Nit: go we need to include the groupId? It's repeated in line 316.

bharatviswa504 replied:
The line is added again to make clear distinguish
Following input topics offsets will be reset to beginning (for consumer group g1)
t1
Following internal topics offsets will be reset to beginning (for consumer group g1
t5

And also consumer group is shown, to provide info to user for this topic offsets are reset for this consumer group.
As most is done to make the output clear

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+        }
+    }
+
+    private void maybeSeekToEnd(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> intermediateTopicPartitions) {
```
Nit: add `final` to parameters

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+
+    private void maybeSeekToEnd(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> intermediateTopicPartitions) {
+
+        final Boolean dryRun = options.has(dryRunOption);
```
Nit: why not use `boolean` ?

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+
+    private void maybeSeekToBeginning(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> inputAndInternalTopicPartitions, List<String> internalTopics) {
+
+        final Boolean dryRun = options.has(dryRunOption);
```
Nit: why not use boolean

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+                client.seekToEnd(intermediateTopicPartitions);
+            } else {
+                System.out.println("Following intermediate topics offsets will be reset to end (for consumer group " + groupId + ")");
+                for (String topic : intermediateTopics) {
```
nit: add `final`

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -104,8 +103,15 @@ public int run(final String[] args, final Properties config) {
+            }
+
+            mayberesetInputAndInternalAndSeekToEndIntermediateTopicOffsets();
+            maybedeleteInternalTopics(zkUtils);
```
nit: `maybeDeleteInternalTopics` and `maybeReset...`

bharatviswa504 replied:
changed.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
+        Boolean dryRun = options.has(dryRunOption);
+        String groupId = options.valueOf(applicationIdOption);
+
+        if (!dryRun) {
```
Not sure why we need this here? If no input topics are specified, we can just return in dry-run case, too?

bharatviswa504 replied:
Added the check to make output clear.
As with out checking input topics, existed or not, we print these input topics offset will be seek to end or beginning.
In dry-run i have taken care that only for topics, which exist print seek to end/begininng. for remaining topics print no action on them.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+            if (notFoundInputTopics.size() > 0) {
+                System.out.println("Following input topics are not found, skipping them");
+                for (String topic : notFoundInputTopics) {
+                    System.out.println("Topic - " + topic);
```
Nit: maybe `("Topic: " + topic)`

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+            if (notFoundIntermediateTopics.size() > 0) {
+                System.out.println("Following intermediate topics are not found, skipping them");
+                for (String topic : notFoundIntermediateTopics) {
+                    System.out.println("Topic - " + topic);
```
Nit: maybe `("Topic: " + topic)`

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
         for (final String topic : inputTopics) {
             if (!allTopics.contains(topic)) {
-                System.err.println("Input topic " + topic + " not found. Skipping.");
+                notFoundInputTopics.add(topic);
```
Why this change?

bharatviswa504 replied:
This is done, to make output clear 
System.out.println("----Dry run displays the actions which will be performed when running Streams Reset Tool----");

As I am printing the above one, and then below printing all dry-run actions, so want to make it clear with description done like this.
Following input topics are not found, skipping them
Topic: t1
Following input topics are not found, skipping them
Topic:t2

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
         for (final String topic : intermediateTopics) {
             if (!allTopics.contains(topic)) {
-                System.err.println("Intermediate topic " + topic + " not found. Skipping.");
+                notFoundIntermediateTopics.add(topic);
```
Why this change?

bharatviswa504 replied:
same as above comment.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
+        final List<String> notFoundInputTopics = new ArrayList<>();
+        final List<String> notFoundIntermediateTopics = new ArrayList<>();
+
+        Boolean dryRun = options.has(dryRunOption);
```
Nit: why not use `boolean`

bharatviswa504 replied:
it can be. Changed it as suggested.

mjsax reviewed (changes requested):
Thanks for the update. Most comments are minor.

One more meta comment: might be better tp use one member variable `boolean dryRun` that is set after `parseArgument()` (L91) instead of all the local `dryRun` vars.

Commit:
addressed review comments

Commit:
addressed review comments

bharatviswa504 commented:
@mjsax Addressed most of the review comments.
Some changes are still left, as they are done to make dry-run output look cleaner.
Below is the dry run output
bin/kafka-streams-application-reset.sh --input-topics streams-file-input,t1,t9 --zookeeper localhost:2181 --dry-run

----Dry run displays the actions which will be performed when running Streams Reset Tool----
Following input topics offsets will be reset to beginning (for consumer group streams-wordcount)
Topic: streams-file-input
Following internal topics offsets will be reset to beginning (for consumer group streams-wordcount)
Topic: streams-wordcount-Counts-repartition
Topic: streams-wordcount-Counts-changelog
Following input topics are not found, skipping them
Topic: t1
Topic: t9
Done.
Following internal/auto-created topics will be Deleted for application streams-wordcount
Topic: streams-wordcount-Counts-repartition
Topic: streams-wordcount-Counts-changelog
Done.

And regarding cleanup for resetting offsets for internal topics, will be done as part of this JIRA 	KAFKA-5210. Once this PR is delivered in.
Please let me know any more changes needed.

bharatviswa504 commented:
retest this please

bharatviswa504 commented:
The test case failures are not related to these changes.

mjsax commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+            if (!dryRun) {
+                client.seekToBeginning(inputAndInternalTopicPartitions);
+            } else {
+                System.out.println("Following input topics offsets will be reset to beginning (for consumer group " + groupId + ")");
```
I understand the motivation and agree it's a good idea. It's only that 

`"Deleting all internal/auto-created topics for application <app.id>."`
and
`"Following internal/auto-created topics will be Deleted for application <app.id>."`

sound super similar. So not sure if we gain much. But it's quite subjective. It's ok if you leave as is. Please fix the typo though `Deleted` -> `deleted`.

mjsax commented:
Call for review @enothereska @dguy

Commit:
removed dryRun check in deletetopic Method

dguy commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -16,10 +16,18 @@
  */
 package kafka.tools;
 
+import java.io.IOException;
```
Why the re-ordering of imports?

bharatviswa504 replied:
I have done re-ordering to follow java standard imports first and then later project specific imports.
Please let me know if you want me to revert it back

dguy replied:
We typically don't re-order imports

bharatviswa504 replied:
reverted back the order to original.

dguy commented on core/src/main/scala/kafka/tools/StreamsResetter.java:
```diff
@@ -219,25 +242,83 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
+
+            maybeSeekToEnd(client, intermediateTopicPartitions);
+
+            if (!dryRun) {
```
could we move this `if(!dryRun)` to before `maybeSeekToBeginning`? Then we don't need the additional checks in `maybeSeekToBeginning` and `maybeSeekToEnd` you could just log the messages you want to log here or in another method.

bharatviswa504 replied:
@dguy as to have a single code path for main logic and dry-run logic it is done this way, 
The intial PR is done like a single check 
if dry-run 
{
}
else {
}

for more info can have a look into @mjsax comments. As the logic between dry-run and actual logic may change if any future changes will be done to this tool, which may show different behavior. So felt it is valid and followed this approach, even if the code does not look cleaner.

dguy reviewed (commented):
Thanks @bharatviswa504 for the patch.
Left a couple of comments, but i'm wondering if there is a way we can do this without having to have `if(!dryRun)` in multiple places? Can we not just have:
```
if (dryRun) {
// print everything that would happend
} else {
  // run the reset
}
```

mjsax commented:
@dguy I don't see a good way to do this without the risk that the "print path" shows something else as the actual "execute path" does. We should ensure that we cannot introduce bug (both paths could diverge if we split them). Does this make sense?

dguy commented:
@mjsax fair enough.

enothereska commented:
LGTM

Commit:
reverted import order

bharatviswa504 commented:
@mjsax @dguy addressed all the review comments.

bharatviswa504 commented:
Test failures are not related to this patch.

mjsax reviewed (approved):
LGTM. Thanks a lot for this patch @bharatviswa504 !

Call for review and merging @guozhangwang

mjsax commented:
Retest this please.

bharatviswa504 commented:
Retest this please.

bharatviswa504 commented:
@guozhangwang could you please help this in reviewing and pushing changes.
As there are there other 2 jira's for streams reset tool, will create PR's once these changes are in.

guozhangwang commented:
Merged to trunk.
